keps/sig-network/20200204-cni-verification-rearchitecture.md

```diff
@@ -0,0 +1,622 @@
+- Have at least 2 nodes.
+- Have pods that are on host network like kube-api-server.
+- Have pods that are on different subnets.
+- Have pods that are on different nodes.
+
+An example test matrix would be (assume each pod has 2 containers, where each uses different ports for connection):
+```
+|      |                  Node-1            |                Node-2              |
+|------|--------------|----------|----------|--------------|----------|----------|
+|      | Host-network | Subnet-A | Subnet-B | Host-network | Subnet-A | Subnet-B |
+|------|--------------|----------|----------|--------------|----------|----------|
+| NS-A | Pod1         | Pod2     | Pod3     | Pod4         | Pod5     | Pod6     |
+| NS-B | Pod7         | Pod8     | Pod9     | Pod10        | Pod11    | Pod12    |
+|      |              |          |          |              |          |          |
+
+```
+One server should be picked from host-network pods and one server from a non-host-network pods (e.g., Pod1 and Pod2 are picked as servers), and all policies should be applied to each servers separately and all the remaining pods start testing their connection based on reachability matrix.
+In summary, with the above matrix, inter/intra-namespace tests on the same node and inter-intra namespace tests on different nodes are demonstrated. Adding the third node and third namespace may provide additional coverage (such as server on ns A whitelisting ns B, but not whitelisting ns C) but having more namespace and nodes may not bring additional value in terms of coverage. 
+
+### Ensuring that large policy stacks evaluate correctly
+
+Right now the coverage of Policy stacks is rudimentary, we may want to test for a large number(i.e. 10) of policies, stacked, depending on whether we think this may be a bug source for providers.
+
+### Ensure NetworkPolicy evaluates correctly regardless of the order of events
+
+The order of Pod/Namespace/NetworkPolicy ADD/UPDATE/DELETE events should not matter while evaluating a NetworkPolicy. In general, CNIs may have different code paths for ADD Pod -> ADD NetworkPolicy order of events versus, ADD NetworkPolicy -> ADD Pod events. At least some tests should focus on the fact that the different order of events for different resources does not affect the evaluation of this NetworkPolicy.
+
+### Generating the reachability matrix on the fly
+Given a network policy and current pods in a cluster, reachability matrices can be generated automatically. This will make adding network policy test easier as they will no longer be defined manually for each test in the code. 
+
+### Consuming network policies as yaml files
+After the above improvement is implemented, since we can generate a reachability matrix on the fly, there will be no need to add network policies to the code. Instead, network policies can be added to a directory as yaml files. 
```

Comment:
you could allow defining NetworkPolicies as YAML now, just inline in the code...

Reply:
Yes we could inline them.  The idea above is that we could literally have YAML files as the *only* input to the tests.  ill clarify that .